### PR TITLE
[ROCm] Optimize Deepseek R1 on MI300X

### PR DESCRIPTION
**Pull request by zhentaocc**
## Motivation

The DeepSeek-V3/R1 model on AMD MI300X (gfx942) was previously limited to fallback code paths for several MLA attention operations because the optimized kernels were gated behind `_use_aiter_gfx95` (gfx950-only). This left significant decode throughput on the table, as gfx942 could not leverage:

1. The fused RoPE + concatenation + KV cache write Triton kernel (`fused_qk_rope_cat_and_cache_mla`), falling back to separate `rotary_emb` + `torch.cat` + attention-managed cache write.
2. Tuned GEMM configurations for DeepSeek-specific matrix shapes on gfx942. resolved in  https://github.com/ROCm/aiter/pull/2024

This PR broadens the optimized paths to all HIP platforms and adds tuned FP8 GEMM configs, improving decode throughput by ~11% on MI300X while maintaining accuracy.

## Modifications

**sglang (`python/sglang/srt/models/deepseek_v2.py`)**:
- **Enable `fused_qk_rope_cat_and_cache_mla` on all HIP platforms**: Changed the guard from `_use_aiter_gfx95` to `_is_hip` (line 1774). This Triton kernel is hardware-agnostic and fuses RoPE application, q/k concatenation, and KV cache write into a single kernel launch.
- **Fix double RoPE application**: Updated the RoPE pre-application condition from `(not _use_aiter or not _is_gfx95_supported or self.use_nsa)` to `(not _use_aiter or not _is_hip or self.use_nsa)` (line 1718). Without this, gfx942 applied RoPE twice (once in `forward_absorb_prepare`, once in the fused kernel), destroying model accuracy (0% on GSM8K).
- **Import `fused_qk_rope_cat_and_cache_mla` for all HIP**: Added a conditional import block under `if _is_hip:` so the fused kernel symbol is available on non-gfx95 HIP platforms.

**sglang (`python/sglang/srt/layers/quantization/fp8_utils.py`)**:
- Added `is_gfx95_supported` import and `_is_gfx95_supported` flag for gfx95-specific quantization paths.

**aiter (GEMM configs)**:
- Added 3 tuned A8W8 blockscale GEMM configuration files for gfx942 targeting DeepSeek-V3/R1 matrix shapes:
  - `N=2112, K=7168` (MLA compressed KV projection)
  - `N=3072, K=1536` (MLA absorbed attention output)
  - `N=7168, K=2048` (MLA decompression)

## Accuracy Tests

```shell
python3 benchmark/gsm8k/bench_sglang.py --host "http://127.0.0.1" --port 8000 --num-shots 8 --num-questions 1319 --parallel 1319 2>&1 | tee /home/zhenchen/projects/sglang/benchmark_results_new/gsm8k_results_fixed.txt
100%|██████████| 1319/1319 [00:55<00:00, 23.74it/s]
Accuracy: 0.954
Invalid: 0.000
Latency: 55.854 s
Output throughput: 2457.117 token/s
```
## Benchmarking and Profiling

Launch server :
```bash
python3 -m sglang.launch_server --model-path /mnt/data/models/deepseek-ai/DeepSeek-V3/ --tp 8 --trust-remote-code
```

### Profiling
```bash
 python3 -m sglang.bench_one_batch_server --model None --base-url http://localhost:8000 --batch-size 256 --input-len 1024 --output-len 1024 --backend sglang --show-report --profile --profile-by-stage --profile-output-dir ./profile_output/main --run-name main
```
Before
Input lens: [1024]. Output lens: [1024].
|   batch size |   input len |   latency (s) |   input throughput (tok/s) |   output throughput (tok/s) | acc length   |   ITL (ms) |   input cost ($/1M) |   output cost ($/1M) | cache hit rate   | profile                                                                          |
|--------------|-------------|---------------|----------------------------|-----------------------------|--------------|------------|---------------------|----------------------|------------------|----------------------------------------------------------------------------------|
|          256 |        1024 |         79.28 |                    21046.9 |                     3923.07 | n/a          |      65.26 |                0.04 |                 0.14 | n/a  

After
Input lens: [1024]. Output lens: [1024].
|   batch size |   input len |   latency (s) |   input throughput (tok/s) |   output throughput (tok/s) | acc length   |   ITL (ms) |   input cost ($/1M) |   output cost ($/1M) | cache hit rate   | profile                                                                               |
|--------------|-------------|---------------|----------------------------|-----------------------------|--------------|------------|---------------------|----------------------|------------------|---------------------------------------------------------------------------------------|
|          256 |        1024 |         72.71 |                      20358 |                     4381.32 | n/a          |      58.43 |                0.04 |                 0.13 | n/a              |

FP8GEMM (qkv_proj, q_b_proj, o_proj)
depends on  https://github.com/ROCm/aiter/pull/2024
Before
image
After
image

ROPE+Cat+Set KV
Before
image
After
image

### Benchmark Results
Before
image
After
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 1]
```diff
@@ -38,6 +38,7 @@
     is_blackwell_supported,
     is_cuda,
     is_flashinfer_available,
+    is_gfx95_supported,
```
Comment (zhentaocc) [inline python/sglang/srt/layers/quantization/fp8_utils.py]:
```suggestion
```

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 2]
```diff
@@ -60,6 +61,8 @@
     from aiter.ops.triton.gemm_a8w8_blockscale import gemm_a8w8_blockscale
 
     aiter_per1x128_quant = get_hip_quant(aiter.QuantType.per_1x128)
+    _is_gfx95_supported = is_gfx95_supported()
```
Comment (zhentaocc) [inline python/sglang/srt/layers/quantization/fp8_utils.py]:
```suggestion
```

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -175,7 +175,12 @@
         fused_qk_rope_cat_and_cache_mla,
         get_dsv3_gemm_output_zero_allocator_size,
     )
-
+if _is_hip:
+    from sglang.srt.layers.rocm_linear_utils import (
+        # aiter_dsv3_router_gemm,
+        fused_qk_rope_cat_and_cache_mla,
+        # get_dsv3_gemm_output_zero_allocator_size,
+    )
```
Comment (zhentaocc) [inline python/sglang/srt/models/deepseek_v2.py]:
```suggestion
if _is_hip:
    from sglang.srt.layers.rocm_linear_utils import fused_qk_rope_cat_and_cache_mla
```

**Review by hubertlu-tw — APPROVED** [review]
LGTM

File: python/sglang/srt/models/deepseek_v2.py [hunk 4]
```diff
@@ -177,7 +177,12 @@
         fused_qk_rope_cat_and_cache_mla,
         get_dsv3_gemm_output_zero_allocator_size,
     )
-
+if _is_hip:
+    from sglang.srt.layers.rocm_linear_utils import (
+        # aiter_dsv3_router_gemm,
+        fused_qk_rope_cat_and_cache_mla,
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/deepseek_v2.py]: This is also a kernel from aiter.

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/deepseek_v2.py (hunk 4 above)
Comment (zhentaocc) [reply python/sglang/srt/models/deepseek_v2.py]: fixed

**Comment by hubertlu-tw** [comment]
> The failure of CI seems to be related to #17780 . Anyone can confirm and fix this? @hubertlu-tw

cc: 
@yctseng0211 @bingxche @HaiShaw

**Review by HaiShaw — APPROVED** [review]

**Review by yichiche — APPROVED** [review]
LGTM

**Review by hubertlu-tw — APPROVED** [review]
LGTM
